src/main/java/io/openliberty/tools/common/plugins/util/PrepareFeatureUtil.java

```diff
@@ -0,0 +1,443 @@
+/**
+ * (C) Copyright IBM Corporation 2021.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.openliberty.tools.common.plugins.util;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.AccessController;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.apache.maven.artifact.versioning.DefaultArtifactVersion;
+import org.apache.commons.io.FileUtils;
+
+
+public abstract class PrepareFeatureUtil extends ServerFeatureUtil {
+
+	private final File installDirectory;
+
+	private String openLibertyVersion;
+
+	public static final String OPEN_LIBERTY_GROUP_ID = "io.openliberty.features";
+	public static final String INSTALL_MAP_ARTIFACT_ID = "install-map";
+	public static final String FEATURES_JSON_ARTIFACT_ID = "features";
+	private static final String MIN_USER_FEATURE_VERSION = "21.0.0.10";
+
+	private File installJarFile;
+	private File jsonFile;
+	
+
+	public PrepareFeatureUtil(File installDirectory, String openLibertyVersion)
+			throws PluginScenarioException, PluginExecutionException {
+		this.installDirectory = installDirectory;
+		this.openLibertyVersion = openLibertyVersion;
+		installJarFile = loadInstallJarFile(installDirectory);
+
+		// check if the openliberty kernel meets min required version 21.0.0.10
+		DefaultArtifactVersion minVersion = new DefaultArtifactVersion(MIN_USER_FEATURE_VERSION);
+		DefaultArtifactVersion version = new DefaultArtifactVersion(openLibertyVersion);
+
+		if (version.compareTo(minVersion) < 0) {
+			throw new PluginScenarioException(
+					"Installing user features on OpenLiberty version "+version+" is not supported. The minimum required version of OpenLiberty for installing user features is "+minVersion+".");
```

Comment:
Will this class get called if the installed runtime is WebSphere Liberty? If so, the message should be more generic to cover both cases (OL and WL).